src/pages/_app.tsx

```diff
@@ -152,3 +157,47 @@ App.getInitialProps = async ({ Component, ctx }) => {
 		host,
 	}
 }
+
+// ------------------------------------------------------------
+// TypeScript helpers for PageComponents' `layout` property
+// @see https://nextjs.org/docs/basic-features/layouts#with-typescript
+// ------------------------------------------------------------
+import type { NextPage } from 'next'
+import type { AppProps, AppContext } from 'next/app'
+
+/**
+ * @usage
+ *
+ * ```typescript
+ * // ./src/page/my-page.tsx
+ * import type { NextPageWithLayout } from "_app"
+ *
+ * const Page: NextPageWithLayout = (props) => {
+ *   // ...
+ * }
+ *
+ * // Page.layout <-- will be typed
+ * ```
+ */
+export type PageWithLayout<P = Record<string, unknown>, IP = P> = NextPage<
+	P,
+	IP
+> & {
+	layout?: React.FC
+}
+
+type ComponentWithLayout = {
+	Component: PageWithLayout & ComponentMaybeWithQuery
+}
+type AppPropsWithLayout = AppProps & ComponentWithLayout
+
+/**
+ * This is our custom param type for `App.getInitialProps`
+ */
+type GetInitialPropsParams = Omit<AppContext, 'Component'> & ComponentWithLayout
+
+/**
+ * This is our custom type for our Next.js `App` component
+ */
+type CustomAppProps = AppPropsWithLayout &
+	Awaited<ReturnType<typeof App['getInitialProps']>>
```

Comment:
To keep this file as short as possible, could we move these to a file under `src/types/`? Maybe `src/types/layouts.ts`?

Reply:
Done